File: src/main/java/net/frozenblock/configurableeverything/biome/util/BiomeConfigUtil.java
package net.frozenblock.configurableeverything.biome.util;

import net.fabricmc.fabric.api.biome.v1.BiomeModification;
import net.fabricmc.fabric.api.biome.v1.BiomeModificationContext;
import net.fabricmc.fabric.api.biome.v1.BiomeModifications;
import net.fabricmc.fabric.api.biome.v1.BiomeSelectors;
import net.fabricmc.fabric.api.biome.v1.ModificationPhase;
import net.frozenblock.configurableeverything.config.BiomeConfig;
import net.frozenblock.configurableeverything.config.MainConfig;
import net.frozenblock.configurableeverything.util.ConfigurableEverythingUtils;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.biome.Biomes;
import net.minecraft.world.level.levelgen.GenerationStep;
import net.minecraft.world.level.levelgen.placement.PlacedFeature;
import java.util.function.Consumer;

public final class BiomeConfigUtil {

	// TODO: fix impl
	public static void init() {
		if (MainConfig.get().biome) {
			var config = BiomeConfig.get();
			var modification = BiomeModifications.create(ConfigurableEverythingUtils.id("feature_modifications"));

			initAddedFeatures(config, modification);
			initRemovedFeatures(config, modification);
			initReplacedFeatures(config, modification);
		}
	}

	private static void initAddedFeatures(BiomeConfig config, BiomeModification modification) {
		var addedFeatures = config.addedFeatures;

		if (addedFeatures != null && addedFeatures.value() != null) {
			for (BiomePlacedFeatureList list : addedFeatures.value()) {
				var biome = list.biome();
				var features = list.features();

				Consumer<BiomeModificationContext> biomeModifyContext = context -> {
					for (DecorationStepPlacedFeature decorationFeature : features) {
						for (ResourceKey<PlacedFeature> placedFeature : decorationFeature.placedFeatures()) {
							context.getGenerationSettings().addFeature(decorationFeature.decoration(), placedFeature);
						}
					}
				};

				biome.ifLeft(tag -> modification.add(ModificationPhase.ADDITIONS, BiomeSelectors.tag(tag), biomeModifyContext));
				biome.ifRight(biomeKey -> modification.add(ModificationPhase.ADDITIONS, BiomeSelectors.includeByKey(biomeKey), biomeModifyContext));
			}
		}
	}

	private static void initRemovedFeatures(BiomeConfig config, BiomeModification modification) {
		var removedFeatures = config.removedFeatures;

		if (removedFeatures != null && removedFeatures.value() != null) {
			for (BiomePlacedFeatureList list : removedFeatures.value()) {
				var biome = list.biome();
				var features = list.features();

				Consumer<BiomeModificationContext> biomeModifyContext = context -> {
					for (DecorationStepPlacedFeature decorationFeature : features) {
						for (ResourceKey<PlacedFeature> placedFeature : decorationFeature.placedFeatures()) {
							context.getGenerationSettings().removeFeature(decorationFeature.decoration(), placedFeature);
						}
					}
				};

				biome.ifLeft(tag -> modification.add(ModificationPhase.REMOVALS, BiomeSelectors.tag(tag), biomeModifyContext));
				biome.ifRight(biomeKey -> modification.add(ModificationPhase.REMOVALS, BiomeSelectors.includeByKey(biomeKey), biomeModifyContext));
			}
		}
	}

	private static void initReplacedFeatures(BiomeConfig config, BiomeModification modification) {
		var replacedFeatures = config.replacedFeatures;

		if (replacedFeatures != null && replacedFeatures.value() != null) {
			for (BiomePlacedFeatureReplacementList list : replacedFeatures.value()) {
				var biome = list.biome();
				var replacements = list.replacements();

				Consumer<BiomeModificationContext> biomeModifyContext = context -> {
					for (PlacedFeatureReplacement replacement : replacements) {
						context.getGenerationSettings().removeFeature(replacement.replacement().decoration(), replacement.original());
						for (ResourceKey<PlacedFeature> placedFeature : replacement.replacement().placedFeatures()) {
							context.getGenerationSettings().addFeature(replacement.replacement().decoration(), placedFeature);
						}
					}
				};

				biome.ifLeft(tag -> modification.add(ModificationPhase.REPLACEMENTS, BiomeSelectors.tag(tag), biomeModifyContext));
				biome.ifRight(biomeKey -> modification.add(ModificationPhase.REPLACEMENTS, BiomeSelectors.includeByKey(biomeKey), biomeModifyContext));
			}
		}
	}
}

File: src/main/java/net/frozenblock/configurableeverything/biome/util/BiomeConfigUtil.kt
package net.frozenblock.configurableeverything.biome.util

import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import net.fabricmc.fabric.api.biome.v1.*
import net.fabricmc.fabric.api.resource.v1.ResourceLoader
import net.frozenblock.configurableeverything.config.BiomeConfig
import net.frozenblock.configurableeverything.config.MainConfig
import net.frozenblock.configurableeverything.util.id
import net.frozenblock.configurableeverything.util.value
import net.frozenblock.lib.sound.api.asImmutable
import net.minecraft.server.packs.PackType
import java.util.function.Consumer

internal object BiomeConfigUtil {

	internal fun init() {
        val config = BiomeConfig.get()
        if (MainConfig.get().biome) {
            val biomeChange = BiomeChange(config.addedFeatures.value, config.removedFeatures.value, config.replacedFeatures.value, config.musicReplacements.value)
            BiomeChanges.add(id("config"), biomeChange)

            val resourceLoader = ResourceLoader.get(PackType.SERVER_DATA)
            resourceLoader.registerReloader(id("biome_changes"), BiomeChanges)
        }
    }

    // should only be run if the config is enabled, since this is only called from the datapack manager
    @JvmStatic
    internal fun applyModifications(changes: Collection<BiomeChange>) = runBlocking {
        val modification: BiomeModification = BiomeModifications.create(id("feature_modifications"))
        changes.forEach { change -> launch {
            change.also { change ->
                // FEATURES
                launch {
                    initAddedFeatures(change, modification)
                }
                launch {
                    initRemovedFeatures(change, modification)
                }
                launch {
                    initReplacedFeatures(change, modification)
                }

                // EFFECTS
                launch {
                    initReplacedMusic(change, modification)
                }
            }
        } }
    }

    private suspend fun initAddedFeatures(change: BiomeChange, modification: BiomeModification) = coroutineScope {
        val addedFeatures = change.addedFeatures
        for (list in addedFeatures) { launch {
            val biome = list.biome
            val features = list.features
            val consumer = Consumer<BiomeModificationContext> { context -> runBlocking {
                for (decorationFeature in features) { launch {
                    val placedFeatures = decorationFeature.placedFeatures
                    for (placedFeature in placedFeatures) { launch {
                        context.generationSettings.addFeature(
                            decorationFeature.decoration,
                            placedFeature
                        )
                    } }
                } }
            } }

            biome.ifLeft {
                modification.add(
                    ModificationPhase.ADDITIONS,
                        BiomeSelectors.includeByKey(it),
                        consumer
                )
            }

            biome.ifRight {
                modification.add(
                    ModificationPhase.ADDITIONS,
                    BiomeSelectors.tag(it),
                    consumer
                )
            }
        } }
    }

    private suspend fun initRemovedFeatures(change: BiomeChange, modification: BiomeModification) = coroutineScope {
        val removedFeatures = change.removedFeatures
        for (list in removedFeatures) { launch {
            val biome = list.biome
            val features = list.features
            val consumer = Consumer<BiomeModificationContext> { context -> runBlocking {
                for (decorationFeature in features) { launch {
                    val placedFeatures = decorationFeature.placedFeatures
                    for (placedFeature in placedFeatures) { launch {
                        context.generationSettings.removeFeature(
                            decorationFeature.decoration,
                            placedFeature
                        )
                    } }
                } }
            } }

            biome.ifLeft {
                modification.add(
                    ModificationPhase.REMOVALS,
                    BiomeSelectors.includeByKey(it),
                    consumer
                )
            }

            biome.ifRight {
                modification.add(
                    ModificationPhase.REMOVALS,
                    BiomeSelectors.tag(it),
                    consumer
                )
            }
        } }
    }

    private suspend fun initReplacedFeatures(change: BiomeChange, modification: BiomeModification) = coroutineScope {
        val replacedFeatures = change.replacedFeatures
        for (list in replacedFeatures) { launch {
            val biome = list.biome
            val replacements = list.replacements
            val consumer: Consumer<BiomeModificationContext> = Consumer<BiomeModificationContext> { context -> runBlocking {
                for (replacement in replacements) { launch {
                    val original = replacement.original
                    val decoration = replacement.replacement.decoration
                    val placedFeatures = replacement.replacement.placedFeatures

                    context.generationSettings.removeFeature(decoration, original)

                    for (placedFeature in placedFeatures) {
                        context.generationSettings.addFeature(decoration, placedFeature)
                    }
                } }
            } }

            biome.ifLeft {
                modification.add(
                    ModificationPhase.REPLACEMENTS,
                    BiomeSelectors.includeByKey(it),
                    consumer
                )
            }

            biome.ifRight {
                modification.add(
                    ModificationPhase.REPLACEMENTS,
                    BiomeSelectors.tag(it),
                    consumer
                )
            }
        } }
    }

    private suspend fun initReplacedMusic(change: BiomeChange, modification: BiomeModification) = coroutineScope {
        val replacedMusic = change.musicReplacements
        for (musicReplacement in replacedMusic) { launch {
            val biome = musicReplacement.biome
            val music = musicReplacement.music.asImmutable ?: return@launch
            val consumer: Consumer<BiomeModificationContext> = Consumer<BiomeModificationContext> { context ->
                //context.attributes.set(EnvironmentAttributes.BACKGROUND_MUSIC, music)
                //context.effects.setMusic(music)
                // todo fix
            }
            biome.ifLeft { modification.add(ModificationPhase.REPLACEMENTS, BiomeSelectors.includeByKey(it), consumer) }
            biome.ifRight { modification.add(ModificationPhase.REPLACEMENTS, BiomeSelectors.tag(it), consumer) }
        } }
    }
}
